8318509: x86 count_positives intrinsic broken for -XX:AVX3Threshold=0

cl4es opened the pull request:
The AVX-512 implementation of count_positives is broken when there are negative values in the tail block, since the len register is repurposed to then re-read as-if it still contained the original value. 

Proposed fix is to restore len from the value in tmp1, which at this point holds the count of bytes in the tail. 

Existing TestCountPositives test is pretty exhaustive and just needs to be run explicitly with AVX3Threshold=0 enabled.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318509](https://bugs.openjdk.org/browse/JDK-8318509): x86 count_positives intrinsic broken for -XX:AVX3Threshold=0 (**Bug** - P4)


### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**) ⚠️ Review applies to [0c6b56b0](https://git.openjdk.org/jdk/pull/16267/files/0c6b56b0c444d1319a26d5f8394a8b0d64f2b0bc)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16267/head:pull/16267` \
`$ git checkout pull/16267`

Update a local copy of the PR: \
`$ git checkout pull/16267` \
`$ git pull https://git.openjdk.org/jdk.git pull/16267/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16267`

View PR using the GUI difftool: \
`$ git pr show -t 16267`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16267.diff">https://git.openjdk.org/jdk/pull/16267.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16267#issuecomment-1770942436)

TobiHartmann reviewed (approved):
Looks good to me.

jatin-bhateja commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -3879,12 +3877,17 @@ void C2_MacroAssembler::count_positives(Register ary1, Register len,
+    {
+      Register tmp3_aliased = len;
+      mov64(tmp3_aliased, 0xFFFFFFFFFFFFFFFF);
+      shlxq(tmp3_aliased, tmp3_aliased, tmp1);
+      notq(tmp3_aliased);
```
We can replace these two instructions with BZHI under BMI2 check.

cl4es replied:
We probably shouldn't do incremental enhancements in this PR. For code enhancements to make much sense we also need a plan to refactor this so that this AVX-512 version can be enabled in production modes (when/if it makes sense).

eme64 commented:
The fix looks reasonable.
But I would suggest that you improve the test as well, or at least add some random example as well (random values in array, random length etc).

cl4es commented:
> But I would suggest that you improve the test as well, or at least add some random example as well (random values in array, random length etc). Currently it is all deterministic, and I'm not sure that the pattern really covers all cases enough.

Ok. I modernized and rewrote the tests to test random array segment lengths and randomly selected negatives. Also added some cases where negatives are only found in the tail.

Also updated TestHasNegatives to run with AVX3Threshold=0 when applicable, and verified both tests still either fail or crash without the patch in this mode.

eme64 commented on test/hotspot/jtreg/compiler/intrinsics/string/TestCountPositives.java:
```diff
@@ -21,88 +21,121 @@
+        int calculated = Helper.StringCodingCountPositives(bytes, off, len);
+        int expected = countPositives(bytes, off, len);
+        if (calculated != expected) {
+            if (expected != len && ng >= 0 && calculated >= 0 && calculated < expected) {
```
Why does this happen? Why does `countPositives` not get the same result here?

cl4es replied:
The `countPositives` intrinsics were adapted from the pre-existing `hasNegatives` intrinsic, and to ease porting over all platforms - including those we don't maintain - we specified `countPositives` so that if there's a negative value at index `idx` then it's acceptable for the intrinsics to return any value from `0` to `idx - 1`. This way intrinsics on all platforms could quickly be adapted over (return either `0` for `hasNegatives == true` or `length` for `hasNegatives == false`) while respective maintainers worked on porting over properly. 

This also allows the intrinsic to do a quick approximation: E.g. if a vectorized loop has checked 128 bytes but then detects a negative in the next block then the intrinsic can opt to return 128 rather than whatever the exact count. The surrounding java code will then do a scalar loop that is roughly as fast as an intrinsic tail routine would be, and with less complexity. I believe the `aarch64` intrinsic is implemented like this.

eme64 replied:
Ok, then you could add such a comment to the code. Optional.

cl4es replied:
I think it's roughly covered by the `@implSpec`:
```
    /**
     * Count the number of leading positive bytes in the range.
     *
     * @implSpec the implementation must return len if there are no negative
     *   bytes in the range. If there are negative bytes, the implementation must return
     *   a value that is less than or equal to the index of the first negative byte
     *   in the range.
     */
    @IntrinsicCandidate
    public static int countPositives(byte[] ba, int off, int len) {
```
.. but of course the historical and current rationale behind that specification might not be immediately clear.

eme64 commented on test/hotspot/jtreg/compiler/intrinsics/string/TestHasNegatives.java:
```diff
@@ -26,81 +26,99 @@
+        assert (len + off < bytes.length);
+        initialize(off, len, maxNegatives);
+        boolean r = Helper.StringCodingHasNegatives(bytes, off, len);
+        if (r ^ ((maxNegatives == 0) ? false : true)) {
```
This condition looks very obscure. Is it not equivalent to: `r ^ maxNegatives != 0`
Or even `r == (maxNegatives == 0)`?

cl4es replied:
The `countPositives` intrinsics were adapted from the pre-existing `hasNegatives` intrinsic, and to ease porting over all platforms - including those we don't maintain - we specified `countPositives` so that if there's a negative value at index `idx` then it's acceptable for the intrinsics to return any value from `0` to `idx - 1`. This way intrinsics on all platforms could quickly be adapted over (return either `0` for `hasNegatives == true` or `length` for `hasNegatives == false`) while respective maintainers worked on porting over properly. 

This also allows the intrinsic to do a quick approximation: E.g. if a vectorized loop has checked 128 bytes but then detects a negative in the next block then the intrinsic can opt to return 128 rather than whatever the exact count. The surrounding java code will then do a scalar loop that is roughly as fast as an intrinsic tail routine would be, and with less complexity. I believe the `aarch64` intrinsic is implemented like this.

eme64 reviewed (approved):
Looks good now, thanks for the work!

cl4es commented:
/integrate
